File: clean.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 20 11:41:06 2016

@author: Zheng Xin
"""

import re


def isEnglish(s):
    try:
        s.encode(encoding='utf-8').decode('ascii')
    except UnicodeDecodeError:
        return False
    else:
        return True

# -- Remove punctuations except: e.g. "'" in "don't"; "-" in "pre-train"
def remove_pun(text):
    if type(text) is str and len(text.split()) > 0:
        text = text.decode('iso-8859-1').encode('utf-8').strip()
        word = text.split()
        __sen = []
        for w in word:
            if w.endswith((',', '.', ';', ':', '?', '!', ')', '>', '\'', '"', '&', '-', '|')):
                w = w[:-1]

            if w.startswith((',', '.', ';', ':', '?', '!', '(', '<', '\'', '"', '&', '-', '|')):
                w = w[1:]

            if w.strip():
                __sen.append(w)

        __msg = ' '.join(__sen)
        __msg = re.sub(r"\(", ' ', __msg)
        __msg = re.sub(r"\)", ' ', __msg)
        __msg = re.sub(r"\[", ' ', __msg)
        __msg = re.sub(r"\]", ' ', __msg)
        __msg = re.sub(r"<", ' ', __msg)
        __msg = re.sub(r">", ' ', __msg)
        __msg = re.sub(r",", ' ', __msg)
        __msg = re.sub(r";", ' ', __msg)
        __msg = re.sub(r"!", ' ', __msg)
        __msg = re.sub(r"$", ' ', __msg)
        __msg = re.sub(r"&", ' ', __msg)
        __msg = re.sub(r"\*", ' ', __msg)
        __msg = re.sub(r"\+", ' ', __msg)
        __msg = re.sub(r"_{2,}", ' ', __msg)
        __msg = re.sub(r"={1,}", ' ', __msg)
        __msg = re.sub(r"-{2,}", ' ', __msg)
        __msg = re.sub(r"~{1,}", ' ', __msg)
        __msg = re.sub(r"\.{2,}", ' ', __msg)
        __msg = re.sub(r"\?{1,}", ' ', __msg)
        __msg = re.sub(r"\{1,}", ' ', __msg)
        __msg = re.sub(r"\"", ' ', __msg)
        __msg = re.sub(r"/", ' ', __msg)
        __msg = re.sub(r"\\", ' ', __msg)
        word = __msg.split()
        __sen = []
        for w in word:
            if w.endswith((',', '.', ';', ':', '?', '!', ')', '>', '\'', '"', '&', '-', '|')):
                w = w[:-1]
            if w.startswith((',', '.', ';', ':', '?', '!', '(', '<', '\'', '"', '&', '-', '|')):
                w = w[1:]
            if w.strip():
                __sen.append(w)

        text = ' '.join(__sen)

    else:
        print text
    return text


# -- Remove urls --
def remove_urls(text):
    urlRegex1 = "^(http|https|ftp|file)://.*"
    urlRegex2 = ".*[/]*.*(.html)"
    left_word = []
    for w in text.split():
        if isEnglish(w) and not (re.match(urlRegex1, w, re.M | re.I)
                                 or re.match(urlRegex2, w, re.M | re.I)):
            left_word.append(w)
    text = " ".join(left_word)
    return text

# -- Remove following automatically generated symbols --
def remove_retweet_sign(text):
    text = re.sub(r"Retweet", "", text)
    text = re.sub(r"retweet", "", text)
    text = re.sub(r"gf", "", text)
    text = re.sub(r"gt", "", text)
    text = re.sub(r"RT", "", text)
    text = re.sub(r"rt", "", text)
    text = text.strip()
    return text


# -- Remove account --
def remove_account(text):
    words = text.split()
    left_word = [word for word in words if "@" not in word]
    text = " ".join(left_word)
    return text


# -- Process various cleaning for input text --
def clean_text(text_iterator, args):
    clean_text = []
    for text in text_iterator:
        if text.strip():
            if args.rm_retweet == 'True':
                text = remove_retweet_sign(text)

            if args.rm_hashtag_sign == 'True':
                text = re.sub(r"#", "", text)

            if args.rm_account_sign == 'True':
                text = re.sub(r"@", "", text)
            elif args.rm_account == 'True':
                text = remove_account(text)

            if args.rm_urls == "True":
                text = remove_urls(text)

            text = remove_pun(text)
            clean_text.append(text)
        else:
            clean_text.append(text)
    return clean_text

File: Clean.java
import java.io.IOException;

/**
 * Created by Zheng Xin on 7/11/2017.
 */
public class Clean {

    static String numRegex = "[0-9]+";
    static String alphaRegex = "[a-zA-Z]+";
    static String NARegex = "[a-zA-Z0-9]+";
    static String urlRegex1 = "^(http|https|ftp|file)://.*";
    static String urlRegex2 = ".*[/]*.*(.html)";
    static String pathRegex = ".*[:\\\\]+.*[>]*.*";
    static String fileRegex = "[.|\\w]*/.*[/]*.*";
    static String ipRegex = "\\W[0-9]+[.]+[0-9]+[.]+[0-9]+[.]+[0-9]+\\W";
    static String hourRegex = "[0-9]+[:]+[0-9]+[:]+[0-9]+.*";
    static String timeRegex = ".*[0-9]+(/|-|.)+[0-9]+(/|-|.)+[0-9]+";

    public static void WordToC(String[] word) throws IOException {
        for(int j = 0; j < word.length; j++){
            word[j]=word[j].trim();
            if (word[j].endsWith(",") || word[j].endsWith(".")
                    ||word[j].endsWith(";") || word[j].endsWith(":")
                    ||word[j].endsWith("?") || word[j].endsWith("!")
                    || word[j].endsWith(")") || word[j].endsWith("'")
                    || word[j].endsWith("\"") || word[j].endsWith(">")
                    || word[j].endsWith("&") || word[j].endsWith("-")) {
                word[j] = word[j].substring(0, word[j].length()-1);
            }
            if(word[j].startsWith(",") || word[j].startsWith(".")
                    ||word[j].startsWith(";") || word[j].startsWith(":")
                    ||word[j].startsWith("?") || word[j].startsWith("!")
                    || word[j].startsWith("(")|| word[j].startsWith("'")
                    || word[j].startsWith("\"") || word[j].startsWith("<")
                    || word[j].startsWith("&") || word[j].startsWith("-")){
                word[j] = word[j].substring(1, word[j].length());
            }
            if(word[j].matches(urlRegex1) || word[j].matches(urlRegex2)
                    || word[j].startsWith(urlRegex1) || word[j].endsWith(urlRegex1)){
                word[j] = "";
            }

            if(word[j].matches("\\W")){
                word[j] = "";
            }

            if(word[j].matches(".*[^a-zA-Z0-9#@,_.]+.*")){
                word[j] = "";
            }

            if(word[j].matches("[a-z]+(,)[a-z]+")){
                String[] sub = word[j].split(",");
                word[j] = "";
                for(String key: sub){
                    word[j] += key + " ";
                }
                word[j] = word[j].trim();
            }

            if(word[j].matches("[a-z]+[.][a-z]+")){
                String[] sub = word[j].split(".");
                word[j] = "";
                for(String key: sub){
                    word[j] += key + " ";
                }
                word[j] = word[j].trim();
            }

            if (word[j].endsWith(",") || word[j].endsWith(".")
                    ||word[j].endsWith(";") || word[j].endsWith(":")
                    ||word[j].endsWith("?") || word[j].endsWith("!")
                    || word[j].endsWith(")") || word[j].endsWith("'")
                    || word[j].endsWith("\"") || word[j].endsWith(">")
                    || word[j].endsWith("&") || word[j].endsWith("-")) {
                word[j] = word[j].substring(0, word[j].length()-1);
            }

            if(word[j].startsWith(",") || word[j].startsWith(".")
                    ||word[j].startsWith(";") || word[j].startsWith(":")
                    ||word[j].startsWith("?") || word[j].startsWith("!")
                    || word[j].startsWith("(")|| word[j].startsWith("'")
                    || word[j].startsWith("\"") || word[j].startsWith("<")
                    || word[j].startsWith("&") || word[j].startsWith("-")){
                word[j] = word[j].substring(1, word[j].length());
            }

            if(word[j].equals("RT") || word[j].equals("rt")|| word[j].equals("amp")
                    || word[j].equals("gt") ){
                word[j]="";
            }

            if(word[j].matches("_")){
                word[j] = "";
            }
        }
    }

    public static String clean(String text) throws Exception {
        text = text.toLowerCase();
        text = text.replaceAll("\\*", " ");
        text = text.replaceAll("\\(", " ");
        text = text.replaceAll("\\)", " ");
        text = text.replaceAll("\\[", " ");
        text = text.replaceAll("\\\\", " ");
        text = text.replaceAll("]", " ");
        text = text.replaceAll("<", " ");
        text = text.replaceAll(">", " ");
        text = text.replaceAll("!", " ");
//        text = text.replaceAll("#", " ");
        text = text.replaceAll("-", " ");
        text = text.replaceAll("$", " ");
        text = text.replaceAll("_{2,}", " ");
        text = text.replaceAll("={2,}", " ");
        text = text.replaceAll("-{2,}", " ");
        text = text.replaceAll("~{2,}", " ");
        text = text.replaceAll("\\.{2,}", " ");
        text = text.replaceAll("\\?{2,}", " ");
//        text = text.replaceAll("\\\"", " ");
        text = text.replaceAll("\"", " ");
        String[] word = text.split("\\s+");
        WordToC(word);
        String text_clean = "";
        for (int i = 0; i < word.length; i++) {
            if (!word[i].isEmpty()) {
                word[i] = word[i].replaceAll("[^a-zA-Z0-9@#,_.]+", " ");
                text_clean += word[i] + " ";
            }
        }
        text_clean = text_clean.trim();
        return text_clean;
    }
}
